clang/include/clang/Basic/LangOptions.h

```diff
@@ -1101,6 +1107,187 @@ inline void FPOptions::applyChanges(FPOptionsOverride FPO) {
+    case AtomicOptionKind::IgnoreDenormalMode:
+      return ignore_denormal_mode;
+    default:
+      llvm_unreachable("Invalid AtomicOptionKind");
+    }
+  }
+
+  void setOption(AtomicOptionKind Kind, bool Value) {
+    switch (Kind) {
+    case AtomicOptionKind::RemoteMemory:
+      remote_memory = Value;
+      break;
+    case AtomicOptionKind::FineGrainedMemory:
+      fine_grained_memory = Value;
+      break;
+    case AtomicOptionKind::IgnoreDenormalMode:
+      ignore_denormal_mode = Value;
+      break;
+    default:
+      llvm_unreachable("Invalid AtomicOptionKind");
+    }
+  }
+
+  LLVM_DUMP_METHOD void dump() const {
+    llvm::errs() << "\n remote_memory: " << remote_memory
+                 << "\n fine_grained_memory: " << fine_grained_memory
+                 << "\n ignore_denormal_mode: " << ignore_denormal_mode << "\n";
+  }
+
+  // The canonical names for each option.
+  static constexpr const char *OptionNames[] = {
+      "remote_memory",       // canonical name for RemoteMemory option.
+      "fine_grained_memory", // canonical name for FineGrainedMemory option.
+      "ignore_denormal_mode" // canonical name for IgnoreDenormalMode option.
+  };
+};
+
+/// Represents overrides for atomic options.
+class AtomicOptionsOverride {
+  AtomicOptions Options;
+  // Bitmask for active overrides.
+  unsigned OverrideMask;
+
+  static constexpr unsigned getMask(AtomicOptionKind Kind) {
+    return 1u << static_cast<unsigned>(Kind);
+  }
+
+public:
+  template <typename Func> static void forEachOption(Func &&func) {
+    for (unsigned I = 0; I < static_cast<unsigned>(
+                                 AtomicOptionKind::LANGOPT_ATOMIC_OPTION_LAST);
+         ++I)
+      func(static_cast<AtomicOptionKind>(I));
+  }
+
+  static constexpr unsigned DefaultMask =
+      (1u << static_cast<unsigned>(
+           AtomicOptionKind::LANGOPT_ATOMIC_OPTION_LAST)) -
+      1;
+
+  AtomicOptionsOverride() : Options(), OverrideMask(0) {}
+  AtomicOptionsOverride(AtomicOptions AO)
+      : Options(AO), OverrideMask(DefaultMask) {}
+  AtomicOptionsOverride(AtomicOptions AO, unsigned Mask)
+      : Options(AO), OverrideMask(Mask) {}
+  AtomicOptionsOverride(const LangOptions &LO) : Options(), OverrideMask(0) {
+    setAtomicOverride(AtomicOptionKind::RemoteMemory, LO.AtomicRemoteMemory);
+    setAtomicOverride(AtomicOptionKind::FineGrainedMemory,
+                      LO.AtomicFineGrainedMemory);
+    setAtomicOverride(AtomicOptionKind::IgnoreDenormalMode,
+                      LO.AtomicIgnoreDenormalMode);
+  }
+  // Parses a token (e.g. "fine_grained_memory" or "no_fine_grained_memory").
+  // If the token begins with "no_", the prefix is stripped and the resulting
+  // string is compared against the canonical name. If it matches, Value is set
+  // to false. Otherwise, if the token exactly matches the canonical name, Value
+  // is true.
+  static std::optional<AtomicOptionKind>
+  parseAtomicOverrideKey(llvm::StringRef Key, bool &Value) {
```

Comment:
Instead of parsing, we should be able to use the `EnumVariadicArgument` option from Attr.td.  I believe there is similar for command-line args with a mild re-factoring.

Reply:
You mean VariadicEnumArgument? Yes using that can avoid the parsing function. will do